### DISCUSSION TITLE
rtl-ais basic advice required

### DISCUSSION THREAD

**Discussion by ajl199** [bareboat-necessities/lysmarine_gen#309] [Q&A]
Sorry in advance, I think part of my problem is that I don't really understand what _should_ be happening.

I have here an rtl-sdr blog stick connected to an AIS antenna, connected to my RPi running BBN. I've done the blacklisting of modules as instructed.

If I run:
    $ rtl_ais -n 
I get a constant stream of AIS messages (AIVDM). So far, so good, I imagine.

If I kill that, and instead do:
  $ systemctl start rtl-ais
  $ ps aux | grep ais
I get:
  /usr/bin/rtl-ais -R on -P 10112 -T
which looks promising.

However:
  $ telnet localhost 10112
gives me four lines of AIVDM only (different every time I do it), and then stops. This may or may not be significant - see me not knowing what to expect.

I've enabled rtl_ais_tcp in Signal K Server, with localhost:10112, TCPclient, data timeout=600, remove null chars=on, as its configured values. However, there's no sign of AIS targets showing up in Freeboard or OpenCPN, or AIVDM messages in OpenCPN's NMEA debug window, where I am seeing other NMEA messages. 

Most of the system config is standard, this is one of the first things I'm trying to get working.

In the SK server logs I do get a repeated 
  TcpProvider:connect ECONNREFUSED 127.0.0.1:30330
which seems to relate to OpenCPN according to the port number shown in the connection and plugin status table of the SK server dashboard. I don't know what that is though - an input to OpenCPN, output from it, plugin, something which requires configuration, or whether it's of any relevance to this.

I'm new to BBN and SignalK, so I don't know whether I'm dealing with a fault of some kind or just haven't set something up right. Any advice would be gladly received; my objective for now is just to get the AIS targets showing up in OpenCPN, but on Friday I'll be trying to connect up my NMEA0183 instruments via this system's MCS board.

**Comment by mgrouch** [comment]
Telnet command should give you continuous stream of AIS messages. Not just 4. 
the rest of the setup looks just fine. 

there is also AIS-catcher. You can try it instead of older rtl_ais.

**Comment by atchik6056** [comment]
Hello
experiencing something similar:

I started rtl_ais to mirror sending data to console:
`
root@lysmarine:/home/signalk/.signalk# /usr/bin/rtl_ais -d 0 -g 49.6 -T -P 10112 -p 86 -n
Found 1 device(s):
  0:  Realtek, RTL2838UHIDIR, SN: rtlsdr

Using device 0: Generic RTL2832U OEM
Edge tuning disabled.
DC filter enabled.
RTL AGC disabled.
Internal AIS decoder enabled.
Buffer size: 163.84 mS
Downsample factor: 64
Low pass: 25000 Hz
Output: 48000 Hz
Found Rafael Micro R820T tuner
Log NMEA sentences to console ON
Tcp listen port 10112
Ais message timeout with 15
Tuner gain set to 49.60 dB.
Tuner error set to 86 ppm.
Tuned to 162000000 Hz.
Sampling at 1600000 S/s.
Allocating 12 zero-copy buffers
!AIVDM,1,1,,B,H3I5;0TTCBD485>675kijp0h622t,0*5D
!AIVDM,1,1,,A,H39vfe0=8U=@Tp4000000000000,0*1C
!AIVDM,1,1,,A,B39``k000?dwF;T0Ra;Q3wWQnDk:,0*60
!AIVDM,1,1,,A,B3I5;0P00?dwE;40Rg;Q3wbQnE2J,0*5D
!AIVDM,1,1,,A,H46CbV4UB1I0000I<jpin0000000,0*18
...
`
it sends continuous stream of data
I can see that it created server on the port 10112
`
user@lysmarine:~ $ sudo netstat -tunlp| grep 10112
tcp        0      0 0.0.0.0:10112           0.0.0.0:*               LISTEN      242747/rtl_ais      
`
however there is nothing in signal k:
`
rtl_ais_tcp	
10/7/2023, 5:39:41 PM: Reconnect 127.0.0.1 10112 retry 0 delay 0
`
signal k channel is configured as
`
   {
      "pipeElements": [
        {
          "type": "providers/simple",
          "options": {
            "logging": false,
            "type": "NMEA0183",
            "subOptions": {
              "validateChecksum": true,
              "type": "tcp",
              "port": "10112",
              "suppress0183event": true,
              "host": "127.0.0.1",
              "removeNulls": false,
              "noDataReceivedTimeout": "0",
              "appendChecksum": false
            }
          }
        }
      ],
      "id": "rtl_ais_tcp",
      "enabled": true
    },
`

netcat does not show anything on port 10112:
`
#nc 127.0.0.1 10112
` 

I also recompiled rtl-ais from https://github.com/dgiardini/rtl-ais, no effect.

PS: the same device and hardware , but on different port (10110 default) rtl-ais was working ok with openplotter, so device is OK

Any idea what is wrong? or just switch to  AIS-catcher?

**Reply by mgrouch** [reply]
Seems rtl_AIS in your case doesn’t send data to tcp socket. You should be able to see data with 

 nc  127.0.0.1   10112

Try starting rtl_AIS without -n option. 

Also in SignalK connection to port 10112 disable checksum checking and enable logging

**Reply by atchik6056** [reply]
this is what I did as described above  sniffing 10112 port with netcat, and running rtl-ais as daemon by systemd no -n.
now after I enable logging I can see in the signalk server log:
Oct 07 21:53:42 TcpProvider:connect ECONNREFUSED 127.0.0.1:10112
Oct 07 21:53:44 TcpProvider:connect ECONNREFUSED ::1:30330
Oct 07 21:53:47 TcpProvider:connect ECONNREFUSED 127.0.0.1:10112
checksum is disabled
it seems some problem with rtl-ais server...
it opens port  and listen to it , but does not send/accept any connection
BTW: ufw is disabled for sake of problem resolution.
I can send output with -n to the pipe and read it by signalk, but it seems to be ugly...
Also I noticed steady high cpu utilization about 20% (pi 4B) by rtl-ais, it seems to be a bit high for such a simple service , not sure...

**Reply by mgrouch** [reply]
I’d try AIS-Catcher. You have it installed too. 

Your hardware seems working as you were able to see AIS messages. 

With netstat -na | grep 10112

do you see any connections established to 10112?

**Reply by atchik6056** [reply]
I did, compiled, installed and running aiscatcher. it is so much better then rtlais, consumes less cpu and provides web server, so I can see it catches a dozen of ships around me.
However signalk does not get anything via udp channel I have created. If I create similar input channel in OpenCPN then I got ais targets but only in opencpn.
 I see nothing on port 10112 in signalk and constant errors on the channel to OpenCPN like
Oct 08 10:27:41 TcpProvider:connect ECONNREFUSED ::1:30330
Oct 08 10:27:46 TcpProvider:connect ECONNREFUSED ::1:30330
and journalctl:
Oct 08 10:29:41 lysmarine signalk-server[41539]: TcpProvider:connect ECONNREFUSED ::1:30330
Oct 08 10:29:46 lysmarine signalk-server[41539]: TcpProvider:connect ECONNREFUSED ::1:30330

also, maybe it is not related, but  signalk cannot upgrade itself from   version 2.1.0 to 2.3.0 

lysmarine version
ser@lysmarine:~ $ uname -a
Linux lysmarine 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux

Note: I have created new channel in signalk to ttyUSB0 RS422 (dhd-tech ftdi) successfully, all works fine.
the issue might be specific to network?

no 10112 is listed in ss -tulpn output

**Comment by atchik6056** [comment]
a bit more:
when I start opencpn , I see
user@lysmarine:~ $ ss -tulpn|grep 10112
udp   UNCONN 0      0            0.0.0.0:10112      0.0.0.0:*    users:(("oexserverd",pid=551187,fd=32),("opencpn",pid=550998,fd=32))

when opencpn is not running , I see nothing, apparently signalk does not fire up connection
user@lysmarine:~ $ ss -tulpn|grep 10112

**Reply by mgrouch** [reply]
you didn’t need to install and compile AIS-catcher because it was already there. 

SignalK as far as I remember had issues with UDP. Can you try AIS-catcher use TCP?

Or at least made sure both are set up to use same either tcp or udp.

**Reply by atchik6056** [reply]
i switched aiscatcher to tcp server -S 10112, configured signalk channel into tcp client 10112 and now I am getting in signalk:
vOct 08 14:47:23 TcpProvider:connect ECONNREFUSED ::1:30330
Oct 08 14:47:23 TcpProvider:connect ECONNREFUSED ::1:10112
though netcat shows constant stream:
user@lysmarine:~ $ nc localhost 10112
!AIVDM,1,1,,A,B31m6wh0<wdw:oT0P<:hCwkQl<0G,0*6C
!AIVDM,1,1,,A,B3P;bCh00?dwCG40RPSQ3wmUoP06,0*1C
!AIVDM,1,1,,A,H3`kCSAALTq<000000000000000,2*21
!AIVDM,1,1,,A,H3`kCSDTDBE5847@3npkh01h0340,0*1B

is there anythings in lysmarine which could be filtering traffic? ufw is disabled

**Reply by mgrouch** [reply]
Even with firewall enabled it should have connected. There is no filter in firewall for connections on local host. 

It’s really weird to me that SignalK can’t connect to TCP. You should contact SignalK dev team on slack 

http://signalk.org/

**Comment by mgrouch** [comment]
Did you specify localhost or 127.0.0.1 in SignalK?

localhost might be resolving to ipv6 address instead of ipv4. 

Put 127.0.0.1 explicitly

**Comment by atchik6056** [comment]
yes, it works with 127.0.0.1
I disable 10112 port in opencpn and still having ais targets
thank you for the hint
though i still have 
Oct 08 16:00:48 TcpProvider:connect ECONNREFUSED 127.0.0.1:30330
even tough i attempted the same trick 127.0.0.1 instead of localhost in the channel for opencpn (defined in bbn by default

**Reply by atchik6056** [reply]
oh, wait a sec, I changed it back to your original 0.0.0.0 and signalk stopped complaining about  
TcpProvider:connect ECONNREFUSED 
can you please comment of this?
and thanks again for your help

PS: lysmarine/bbn works and does what I want so far 
even connecting to cellular network via quectel usb modem, it is my major routing device now

**Reply by mgrouch** [reply]
Binding of server to 0.0.0.0 means 
Bind to all network interfaces in case of you have multiple networks (ex WiFi and wired Ethernet)

**Reply by mgrouch** [reply]
Port 30330 is only used when you activate routing to waypoint to send commands to pypilot. So warning about not connecting to 30330 is normal and you simply can ignore it.

**Reply by atchik6056** [reply]
thank you very much for your quick and helpful answers
all good now
